### ISSUE TITLE
Kellnr calculates wrong checksum [CRITICAL]

### ISSUE THREAD

**Issue by alexthe2** [kellnr/kellnr#311]
Since last week (there was no particular event), kellnr started calculating the wrong checksums for the crates pushed to it.

This results in cargo not being able to verify checksums.
```
failed to verify the checksum of `v12-data v1.17.1 (registry `kellnr`)
```

As an example crate:
```sql
sqlite> select * from crate_index where name="v12-data" AND vers="1.17.1";
id|name|vers|deps|cksum|features|yanked|links|v|crate_fk
35|v12-data|1.17.1|...|90392cdc01079ee500391dd0fd409059158b83b170567361d2e3552b76450f6c|{}|0||1|5
```
The checksum here is not the checksum that would occur from running `sha256sum`, when we did a `UPDATE crate_index ...` with the result of sha256, that particular crate started correctly working (with cargo).

We're happy to help with resolving this issue asap, as it's a criticial bug for us, but would need some guidance as to where this issue could be originating from.

We're running the latest kellnr 5.2.1, with rust 1.77.2 on the buildserver (Jenkins) that pushes to kellnr, and locally 1.75+

**Comment by secana** [comment]
Hi @alexthe2! Thanks for reporting the issue. Does this happen only with one specific crate or under specific circumstances? I'll try to reproduce the issue.

**Comment by alexthe2** [comment]
It happens with two crates for us (from what I can see right now), called v12-data and v12-terra_converters, `v12-terra_converters` is depndent on data

**Comment by secana** [comment]
Can you provide logs from Kellnr in the "trace" mode, when the issue occurs? I try to replicate the wrongly computed hash, but so far without any success.

**Comment by alexthe2** [comment]
should I set `KELLNR_LOG__LEVEL` or `KELLNR_LOG__LEVEL_WEB_SERVER`?

**Comment by secana** [comment]
KELLNR_LOG__LEVEL should be enough.

**Comment by alexthe2** [comment]
it's not producing any logs 😢 , I verfied that the level is really set to trace (updated helm chart, and deleted pod to force restart)

**Comment by asymmetry** [comment]
Hi @secana, we are experiencing the same issue. I could provide one more data point.

Compare the sha256sum of a corrupted crate in the db and the actual value:
In db:
![image](https://github.com/kellnr/kellnr/assets/679529/54901db6-d31c-4627-bee2-761f54e87647)
Actual:
![image](https://github.com/kellnr/kellnr/assets/679529/4e46f11b-8602-4c9f-b2d4-fefbd9e375c0)

And I am already running with KELLNR_REGISTRY__CACHE_SIZE=0
![image](https://github.com/kellnr/kellnr/assets/679529/4f41d969-acdf-42ec-a178-3156d2a1808d)

I am running Kellnr 5.2.2 with the released docker image.

Thanks!

**Comment by asymmetry** [comment]
If I delete the corrupted version from the web ui as admin user, restart the docker container, and then publish the same crate again, it could fix the problem.

**Comment by alexthe2** [comment]
Fun story for us, if you hand fix the cksum in SQLite, it might solve the
problem for the next version.

The reason I say might is it solved it for one crate but not for all of them

**Comment by secana** [comment]
Thanks for the input. So far, the issue seems to be the computation of the sha256, as the crate itself seems to be fine on disk. I released a debug version of kellnr with much more debug output for the specific issue. Would you be so kind and try to run it and provide the logs here?

Kellnr version: 5.2.3-debug-311
Helm chart version: 3.2.3-debug-311

All logs are in the level `debug` and prefixed with `#311` to be easily identifiable. For the debug version, I added an additional crate to compute the sha256 and be able to compare it with the current implementation. Hopefully the debug output shows us the right direction to finally find and fix the issue.

**Comment by asymmetry** [comment]
Thanks @secana! I have deployed this test version with logs enabled. I tried to publish a test crate 10 times (with slight modification each time) and the logs looks good to me. My team will continue to use this version and I will show the logs here if the issue happens again.

Thanks for the help!

**Comment by secana** [comment]
New version of Kellnr 5.3.2 is out with improved SHA256 computation. This error should be fixed with that version.

**Comment by alexthe2** [comment]
Also a finishing update from our side, we got the debug version running two days ago, no issues yet, we'll now switch to the new 5.3.2

**Comment by asymmetry** [comment]
Hi @secana, my team have the debug version running for almost 10 days and we captured this issue again. Here is the log:
![image](https://github.com/kellnr/kellnr/assets/679529/a39a0130-6cdc-4e10-89ca-f9035c0c73c6)
It seems that the error is happenned when reading the crate saved to disk back into memory for sha256 calculation, it only reads 4096 bytes.
However it seems you have already switched to use the in-memory data to calcualte checksum so I think this is not going to happen again. We will switch to the new 5.3.2.
Thanks a log for the help!

**Comment by asymmetry** [comment]
Hi @secana, sorry to keep posting on a closed issue, but I did some study for the actual root cause.

I was able to reproduce the reading issue by this minimal reproducible example:

```rust
use std::path::Path;

use anyhow::Result;
use tokio::{
    fs::File,
    io::{AsyncReadExt, AsyncWriteExt},
};

#[tokio::main]
async fn main() -> Result<()> {
    let path = Path::new("test/source.crate");

    let mut file = File::open(path).await?;

    let mut source = Vec::new();
    let _ = file.read_to_end(&mut source).await?;

    let source_length = source.len();

    println!("source length: {}", source_length);

    let mut test = 1;

    loop {
        let path = Path::new("test/test.crate");

        let mut file = File::create(&path).await?;
        file.write_all(&source).await?;

        // if we comment this line, we could reproduce the read issue
        file.flush().await?;

        let mut buf: Vec<u8> = vec![];

        let mut file = File::open(&path).await?;
        let bytes_read = file.read_to_end(&mut buf).await?;
        let real_length = file.metadata().await?.len();

        if bytes_read != source_length || real_length != source_length as u64 {
            println!(
                "test {}, bytes read: {}, real length: {}",
                test, bytes_read, real_length
            );
            break;
        }

        test += 1;
    }

    Ok(())
}
```

As shown in the comment, it seems that we should flush the io buffer after the write_all() call, otherwise we will trigger this issue occasionally. There is an useful comment from the maintainer of Tokio: https://users.rust-lang.org/t/async-write-all-sometimes-silently-fails/68195/4.

Not sure if it is still valid for the new version but maybe a flush() call is helpful.

**Comment by secana** [comment]
@asymmetry Thank you so much! With the in-memory computation I build around the issue but knowing what went wrong initially gives me piece of mind. I'll check if I use the same code somewhere else in kellnr and make sure that I flush the buffer.

**Comment by Christiaan676** [comment]
@secana Hi, looking at the code for the package download I think it has the same issue. In `cratesio_api.rs` the function `download()` will basically do the following three actions in order:

1. Get file  path from the crates_storage `file_path = crate_storage.crate_path`;
2. Check if it exists, if not download it and store it in the `crate_storage.add_bin_package`, that will use `write_all` to write the crate to disk;
3. Read the contents from disk `crate_storage.get_file`.

For the initial download this means there are two IO actions. Write to disk, and than read back and store in cache. I Would suggest to just keep the data in memory here.

**Comment by plied** [comment]
You guys mean it's fixed in version `5.2.3` i guess. Version `5.3.2` does not exist yet.

**Comment by secana** [comment]
Yes, you are right. That' a typo. 5.2.3 is the right version.
